src/main/java/org/openjdk/jextract/impl/NameMangler.java

```diff
@@ -196,14 +201,45 @@ public Void visitTypedef(Declaration.Typedef typedef, Declaration parent) {
            JavaFunctionalInterfaceName.with(typedef, javaName);
            functionTypeDefNames.put(typedef.type(), javaName);
         }
+
+        // handle if this typedef is of a struct/union/enum etc.
```

Comment:
Note (here and elsewhere): we need to recurse to nested declarations _last_. Otherwise a nested declaration can take the "good" name of the current declaration (which will then be mangled to `<name>$0`).